### PR TITLE
feat: add k8s Job support to verify and status check

### LINKED ISSUE #8139: [FR] Add support for `skaffold verify` to run tests in cluster as well (not just in containers locally)
Issue tracks feature request to add support for `skaffold verify` to run tests in cluster as well (not just in containers locally)

### PR DESCRIPTION
**Pull request by aaron-prindle**
fixes #8139 

NOTE: now updated to meet backwards compatible schema vs updated schema

### CODE REVIEW

File: pkg/skaffold/verify/k8sjob/verify.go [hunk 1]
```diff
@@ -0,0 +1,461 @@
...
+// NewVerifier returns a new Verifier for a VerifyConfig filled
+// with the needed configuration for `kubectl apply`
+func NewVerifier(ctx context.Context, cfg kubectl.Config, labeller *label.DefaultLabeller, v *latest.KubernetesClusterVerifier, artifacts []*latest.Artifact) (*Verifier, error) {
+	defaultNamespace := ""
+	b, err := (&util.Commander{}).RunCmdOut(context.Background(), exec.Command("kubectl", "config", "view", "--minify", "-o", "jsonpath='{..namespace}'"))
+	if err == nil {
+		defaultNamespace = strings.Trim(string(b), "'")
+		if defaultNamespace == "default" {
+			defaultNamespace = ""
+		}
```
Comment (renzodavid9) [inline pkg/skaffold/verify/k8sjob/verify.go:85]: According to comment in line 95, IIUC, I think we don't need this?

File: pkg/skaffold/verify/k8sjob/verify.go [hunk 2]
```diff
@@ -0,0 +1,461 @@
...
+			return nil, err
+		}
+	}
+	kubectl := kubectl.NewCLI(cfg, v.Flags, defaultNamespace)
+	// default namespace must be "default" not "" when used to create and stream logs from Job(s)
+	if defaultNamespace == "" {
+		defaultNamespace = "default"
+	}
+	tracker := tracker.NewContainerTracker()
+	logger, err := k8sjoblogger.NewLogger(ctx, tracker, labeller, kubectl.KubeContext)
```
Comment (renzodavid9) [inline pkg/skaffold/verify/k8sjob/verify.go:93]: Right now, the `k8sjoblogger.NewLogger`, returns always nil as error, is it possible to remove the returned error?

File: pkg/skaffold/verify/k8sjob/verify.go [hunk 3]
```diff
@@ -0,0 +1,461 @@
...
+
+func (v *Verifier) RegisterLocalImages(images []graph.Artifact) {
+	v.localImages = images
+}
+
+func (v *Verifier) TrackBuildArtifacts(artifacts []graph.Artifact) {
+	v.logger.RegisterArtifacts(artifacts)
+}
+
+// Verify executes specified artifacts by creating containers in the local docker daemon
```
Comment (renzodavid9) [inline pkg/skaffold/verify/k8sjob/verify.go:136]: This looks like the description for the docker verifier, here it is a k8s job, not a container in the local docker daemon.

File: pkg/skaffold/verify/k8sjob/verify.go [hunk 4]
```diff
@@ -0,0 +1,461 @@
...
+				builds = append(builds, graph.Artifact{
+					ImageName: tc.Container.Image,
+					Tag:       b.Tag,
+				})
+				nTC.Container.Image = na.Tag
+				break
+			}
+		}
+		if !foundArtifact {
+			na = graph.Artifact{
```
Comment (renzodavid9) [inline pkg/skaffold/verify/k8sjob/verify.go:182]: Not sure if I understood the `na` variable correctly, but does it looks like we are not using it in here?

File: pkg/skaffold/verify/k8sjob/verify.go [hunk 5]
```diff
@@ -0,0 +1,461 @@
...
+	//   detecting the API type we are dealing with, for
+	//   accurate type casting later.
+	obj, groupVersionKind, err := decoder.Decode(
+		[]byte(resourceYAML),
+		nil,
+		nil)
+	if err != nil {
+		return nil, err
+	}
+	// Only process Deployments for now
```
Comment (renzodavid9) [inline pkg/skaffold/verify/k8sjob/verify.go:399]: Should be `Jobs`?

File: pkg/skaffold/verify/k8sjob/verify.go (hunk 2 above)
Comment (aaron-prindle) [reply pkg/skaffold/verify/k8sjob/verify.go]: done

File: pkg/skaffold/verify/k8sjob/verify.go (hunk 3 above)
Comment (aaron-prindle) [reply pkg/skaffold/verify/k8sjob/verify.go]: done

File: pkg/skaffold/verify/k8sjob/verify.go (hunk 5 above)
Comment (aaron-prindle) [reply pkg/skaffold/verify/k8sjob/verify.go:397]: done

File: pkg/skaffold/verify/k8sjob/verify.go (hunk 4 above)
Comment (aaron-prindle) [reply pkg/skaffold/verify/k8sjob/verify.go]: done, removed na entirely

File: pkg/skaffold/verify/k8sjob/verify.go (hunk 1 above)
Comment (aaron-prindle) [reply pkg/skaffold/verify/k8sjob/verify.go:85]: Currently code has changed, not sure what comment you are referring to.  While a bit convoluted, currently this is neccessary IIUC as this value when used when creating the kubectl CLI must have "" and not "default"
```
	kubectl := kubectl.NewCLI(cfg, latest.KubectlFlags{}, defaultNamespace)
```
and then in the `Verify` code it must be "default" and not ""

Comment (renzodavid9) [reply pkg/skaffold/verify/k8sjob/verify.go:85]: You're right, didn't notice. the use in `kubectl.NewCLI`, my bad.

File: pkg/skaffold/verify/verify_mux.go [hunk 6]
```diff
@@ -52,71 +50,56 @@ func (m VerifierMux) GetVerifiers() []Verifier {
...
-				return err
-			}
-		}
+
+		// if runHooks || m.iterativeStatusCheck {
+		// 	if err := verifier.GetStatusMonitor().Check(ctx, w); err != nil {
+		// 		endTrace(instrumentation.TraceEndError(err))
+		// 		return err
+		// 	}
+		// }
```
Comment (renzodavid9) [inline pkg/skaffold/verify/verify_mux.go:100]: Can we remove this commented code?

File: pkg/skaffold/verify/verify_mux.go [hunk 7]
```diff
@@ -52,71 +50,56 @@ func (m VerifierMux) GetVerifiers() []Verifier {
...
 		if ok {
-			runHooks = deployHooks.HasRunnableHooks()
+			runHooks = verifyHooks.HasRunnableHooks()
 		}
 		if runHooks {
-			if err := deployHooks.PreDeployHooks(ctx, w); err != nil {
+			if err := verifyHooks.PreDeployHooks(ctx, w); err != nil {
 				return err
 			}
 		}
```
Comment (renzodavid9) [inline pkg/skaffold/verify/verify_mux.go:87]: Do we want to support hooks for verify in the future? Or is it safe to delete this code?

Comment (aaron-prindle) [reply pkg/skaffold/verify/verify_mux.go:87]: We can had this in the future, doesn't make sense to have unused code here (there is no schema for hooks in verify currently).

Comment (aaron-prindle) [reply pkg/skaffold/verify/verify_mux.go:87]: Removed this for now, doesn't make sense to have unused stub/dead code.  We do want to add this in the future, created https://github.com/GoogleContainerTools/skaffold/issues/8472 to track this

File: pkg/skaffold/verify/verify_mux.go (hunk 6 above)
Comment (aaron-prindle) [reply pkg/skaffold/verify/verify_mux.go]: Whoops, missed this.  Done now, thanks!\

**Review by renzodavid9 — APPROVED** [review]

**Pull request merged**